tests/util_bucket.sh

```diff
@@ -0,0 +1,304 @@
+  fi
+
+  if [[ $LOG_LEVEL_INT -ge 5 ]]; then
+    if ! log_bucket_policy "$1"; then
+      log 2 "error logging bucket policy"
+      return 1
+    fi
+  fi
+
+  if ! check_object_lock_config "$1"; then
+    log 2 "error checking object lock config"
+    return 1
+  fi
+
+  if [[ "$DIRECT" != "true" ]] && ! add_governance_bypass_policy "$1"; then
+    log 2 "error adding governance bypass policy"
+    return 1
+  fi
+
+  if ! list_and_delete_objects "$1"; then
+    log 2 "error listing and deleting objects"
+    return 1
+  fi
+
+  #run check_ownership_rule_and_reset_acl "$1"
+  #assert_success "error checking ownership rule and resetting acl"
+
+  # shellcheck disable=SC2154
+  if [[ $lock_config_exists == true ]] && ! put_object_lock_configuration_disabled "$1"; then
+    log 2 "error disabling object lock config"
+    return 1
+  fi
+  #if ! put_bucket_versioning "s3api" "$1" "Suspended"; then
+  #  log 2 "error suspending bucket versioning"
+  #  return 1
+  #fi
+
+  #if ! change_bucket_owner "$AWS_ACCESS_KEY_ID" "$AWS_SECRET_ACCESS_KEY" "$1" "$AWS_ACCESS_KEY_ID"; then
+  #  log 2 "error changing bucket owner back to root"
+  #  return 1
+  #fi
+}
+
+# params:  bucket name
+# return 0 if able to delete recursively, 1 if not
+delete_bucket_recursive_s3api() {
+  log 6 "delete_bucket_recursive_s3api"
+  if [ $# -ne 1 ]; then
+    log 2 "'delete_bucket_recursive_s3api' requires bucket name"
+    return 1
+  fi
+
+  if ! clear_bucket_s3api "$1"; then
+    log 2 "error clearing bucket (s3api)"
+    return 1
+  fi
+
+  if ! delete_bucket 's3api' "$1"; then
+    log 2 "error deleting bucket"
+    return 1
+  fi
+  return 0
+}
+
+# params: client, bucket name
+# return 0 on success, 1 on error
+delete_bucket_contents() {
+  log 6 "delete_bucket_contents"
+  if [ $# -ne 2 ]; then
+    log 2 "'delete_bucket_contents' requires client, bucket name"
+    return 1
+  fi
+
+  local exit_code=0
+  local error
+  if [[ $1 == "aws" ]] || [[ $1 == 's3api' ]]; then
+    if ! clear_bucket_s3api "$2"; then
+      log 2 "error clearing bucket (s3api)"
+      return 1
+    fi
+  elif [[ $1 == "s3cmd" ]]; then
+    delete_bucket_recursive "s3cmd" "$1"
+  elif [[ $1 == "mc" ]]; then
+    delete_bucket_recursive "mc" "$1"
+  elif [[ $1 == "s3" ]]; then
+    delete_bucket_recursive "s3" "$1"
+  else
+    log 2 "unrecognized client: '$1'"
+    return 1
+  fi
+  return 0
+}
+
+# check if bucket exists
+# param:  bucket name
+# return 0 for true, 1 for false, 2 for error
+bucket_exists() {
+  if [ $# -ne 2 ]; then
+    log 2 "bucket_exists command requires client, bucket name"
+    return 2
+  fi
+  local exists=0
+  head_bucket "$1" "$2" || exists=$?
+  # shellcheck disable=SC2181
+  if [ $exists -ne 0 ] && [ $exists -ne 1 ]; then
+    log 2 "unexpected error checking if bucket exists"
+    return 2
+  fi
+  if [ $exists -eq 0 ]; then
+    return 0
+  fi
+  return 1
+}
+
+# params:  client, bucket name
+# return 0 for success, 1 for error
+delete_bucket_or_contents() {
```

Comment:
I'm a little confused as to this method - why isn't it two methods?  One for deleting only bucket contents and one for deleting the bucket itself?  Considering that it is mainly just a wrapper with an variable check it just seems weird to me.

Reply:
It was just to combine operations for static and non-static bucket cleanup.  For both, occasionally you need to delete individual objects due to WORM protection.  And just to avoid putting the static check in each test.  If you think there's a better way to do it though, let me know.